sql: column with OID type breaks ReplaceIDsWithFQNames

**Describe the problem**

The latest version of CockroachDB (f87728fee) crashes when executing the following query:

```sql
# poc
CREATE TABLE v0 (c1 OID); 
ALTER TABLE v0 ADD CHECK (c1 < 4294967295); 
```


**To Reproduce**

Here is the detail steps to reproduce the bug.

1. In system Ubuntu 20.04, download the CockroachDB source code from the github source.
2. Use the latest version of the CockroachDB code (tested version: f87728fee)
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment.
5. Observe the crashing and log the crashing information. 

**Expected behavior**
The `ALTER TABLE ADD CHECK` statement should be executed correctly. 

**Additional data / screenshots**

Here is the crashing stack trace:

```
demo@127.0.0.1:26257/movr> CREATE TABLE v0 (c1 OID);
CREATE TABLE


Time: 25ms total (execution 25ms / network 0ms)

demo@127.0.0.1:26257/movr> ALTER TABLE v0 ADD CHECK (c1 < 4294967295);
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x17f94ac]

goroutine 4846 [running]:
github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr.ReplaceIDsWithFQNames.func1({0x7975798, 0x40031563e0})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr/column.go:337 +0x4c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*simpleVisitor).VisitPre(0x40004c1260, {0x7975798, 0x40031563e0})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:1713 +0x48
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.WalkExpr({0x79318e8, 0x40004c1260}, {0x7975798?, 0x40031563e0})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:825 +0x44
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Walk(0x40016e4910, {0x79318e8, 0x40004c1260})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:192 +0x64
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.WalkExpr({0x79318e8, 0x40004c1260}, {0x79753a8?, 0x40016e4910})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:828 +0x5c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.SimpleVisit({0x79753a8, 0x40016e4910}, 0x4003156400)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:1732 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr.ReplaceIDsWithFQNames({0x7959568?, 0x4004be3240}, {0x79753a8, 0x40016e4910}, 0x4001540620)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr/column.go:357 +0xb0
github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr.formatExprForDisplayImpl({0x7959568, 0x4004be3240}, {0x7a12360?, 0x4001b7f900?}, {0x4001fcfbc0?, 0x4e0310?}, 0x4004dc0638?, 0x4004446f00, 0xa46ed00?, 0x0)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr/expr.go:259 +0x68
github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr.FormatExprForDisplay(...)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/schemaexpr/expr.go:211
github.com/cockroachdb/cockroach/pkg/sql.validateCheckExpr({0x7959568?, 0x4004be3240?}, 0x4000100000?, 0x40036775a0?, 0x0?, {0x4001fcfbc0?, 0x15?}, 0x4001b7f900?, {0x79a7e70?, 0x4000d2f500?})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/check.go:55 +0x74
github.com/cockroachdb/cockroach/pkg/sql.validateCheckInTxn.func1()
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:2715 +0x44
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).WithSyntheticDescriptors(0x4a11ae0?, {0x40036775a0?, 0x0?, 0x0?}, 0x0?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:113 +0x84
github.com/cockroachdb/cockroach/pkg/sql.validateCheckInTxn({0x7959568?, 0x4004be3240}, 0x4001540620, 0x4004446f00, 0x4001b7f900, 0x40015f6420, {0x79a7e70?, 0x4000d2f500}, {0x4001fcfbc0, 0x15})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:2712 +0x248
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).validateConstraints.func2.1({0x7959568, 0x4004be3240}, 0x40015f6420, 0x4002a03c00, {0x79a7e70, 0x4000d2f500})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:791 +0x280
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).makeFixedTimestampRunner.func1.1({0x7959568, 0x4004be3240}, 0x4004be3240?, 0x4003156340?, 0x400380e140?, {0x79a7e70, 0x4000d2f500})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:165 +0x120
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).fixedTimestampTxnWithExecutor.func1({0x7959568, 0x4004be3240}, 0x4003b15920?, 0x0?, 0x0?, {0x79a7e70, 0x4000d2f500})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:221 +0xa4
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).txnWithExecutor.func1({0x7959568?, 0x4004be3240?}, 0x4004be3240?, 0x40015f6420?, {0x79a7e70?, 0x4000d2f500?})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:2485 +0x48
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutorFactory).DescsTxnWithExecutor.func2({0x7959568, 0x4004be3240}, 0x7959610?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1419 +0x194
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x7959568?, 0x4004be3240?}, 0x79a01c0?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:960 +0x2c
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0x40015f6420, {0x7959568, 0x4004be3240}, 0x4003c0fc48)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:926 +0x70
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x7959568, 0x4004be3240}, 0x7?, 0x45267c?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:959 +0x58
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0xffffaa60ba68?, {0x7959568, 0x4004be3240}, 0x4be3300?, 0x40?, 0x0?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:922 +0x90
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0x38?, {0x7959568?, 0x4004be3240?}, 0x4004600c70?)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:901 +0x30
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutorFactory).DescsTxnWithExecutor(0x4001795a40, {0x7959568, 0x4004be3240}, 0x30?, 0x4004446f00, 0x40004c10f8, {0x0?, 0x4687120?, 0xffffaa60ba01?})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:1414 +0x168
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).txnWithExecutor(0x4001780200, {0x7959568, 0x4004be3240}, 0x4003156320)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/schema_changer.go:2481 +0xf8
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).fixedTimestampTxnWithExecutor(0x4000100000?, {0x7959568, 0x4004be3240}, {0x1?, 0x2965ed8?, 0x40?}, 0x4003471050)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:214 +0xa0
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).makeFixedTimestampRunner.func1({0x7959568, 0x4004be3240}, 0x4003156300)
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:156 +0xe4
github.com/cockroachdb/cockroach/pkg/sql.(*SchemaChanger).validateConstraints.func2({0x7959568, 0x4004be3240})
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/backfill.go:777 +0x104
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x2c
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:74 +0x5c
created by golang.org/x/sync/errgroup.(*Group).Go
        /home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:71 +0xa4
```

**Environment:**
 - CockroachDB version: v23.1.0-alpha
 - Server OS: Ubuntu 20.04 LTS
 - Client app: CockroachDB demo command. (`./cockroach demo`)
 - Detailed CockroachDB version output:

```
cockroach version details:
Build Tag:        v23.1.0-alpha.00000000-200-gf87728fee6-dirty
Build Time:       2022/11/10 19:55:22
Distribution:     CCL
Platform:         linux arm64 (aarch64-linux-gnu)
Go Version:       go1.19
C Compiler:       gcc 9.4.0
Build Commit ID:  f87728fee6d16e2a352ac76e7ae9225930032ce6
Build Type:       development
```


**Additional context**

The problem is reproducible on all versions from `v22.2.*` and `v22.1.*` and `v21.2.*`. I did not test the versions before `v21.2.*`. 

Jira issue: CRDB-21388

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -793,6 +793,12 @@ func (sc *SchemaChanger) validateConstraints(
 				resolver := descs.NewDistSQLTypeResolver(collection, txn.KV())
 				semaCtx := tree.MakeSemaContext()
 				semaCtx.TypeResolver = &resolver
+				semaCtx.NameResolver = NewSkippingCacheSchemaResolver(
+					txn.Descriptors(),
+					sessiondata.NewStack(NewFakeSessionData(&sc.settings.SV, "validate constraint")),
+					txn.KV(),
+					nil, /* authAccessor */
+				)
 				semaCtx.FunctionResolver = descs.NewDistSQLFunctionResolver(collection, txn.KV())
 				// TODO (rohany): When to release this? As of now this is only going to get released
 				//  after the check is validated.

--- a/pkg/sql/logictest/testdata/logic_test/check_constraints
+++ b/pkg/sql/logictest/testdata/logic_test/check_constraints
@@ -393,3 +393,14 @@ INSERT INTO t67100b VALUES (1)
 
 statement error failed to satisfy CHECK constraint \(true AND \(0:::INT8 > 1:::INT8\)\)
 UPSERT INTO t67100b VALUES (1)
+
+subtest regression_91697
+
+statement ok
+CREATE TABLE t_91697(a OID);
+
+statement ok
+ALTER TABLE t_91697 ADD CHECK (a < 123);
+
+statement error pgcode 23514 pq: failed to satisfy CHECK constraint \(a < 'public.t67100a'::REGCLASS\)
+INSERT INTO t_91697 VALUES (321);

